### PR TITLE
Calendar: implement new showWeekNumbers option

### PR DESCRIPTION
**Pull request by Zedwag**

### CODE REVIEW

File: js/core/utils/date.js [hunk 1]
```diff
@@ -672,6 +681,7 @@ const dateUtils = {
     getLastMonthDate: getLastMonthDate,
     getFirstMonthDate: getFirstMonthDate,
     getFirstWeekDate: getFirstWeekDate,
+    getWeekNumber: getWeekNumber,
```
Comment (marker-dao) [inline js/core/utils/date.js:683]: Can we use a short syntax like this?

```
const dateUtils = {
     getWeekNumber,
}
```

Comment (Zedwag) [reply js/core/utils/date.js:683]: nope, untill we export one object. I guess we can refactor these utils to seperately export all function we need, but not sure it's we should do now, as long as other teams can be "tied" with current syntax

File: js/ui/calendar/ui.calendar.base_view.js [hunk 2]
```diff
@@ -85,6 +85,9 @@ const BaseView = Widget.inherit({
             for(let colIndex = 0, colCount = this.option('colCount'); colIndex < colCount; colIndex++) {
                 this._renderCell(rowData, colIndex);
             }
+            if(this._shouldRenderWeekColumn()) {
```
Comment (ksercs) [inline js/ui/calendar/ui.calendar.base_view.js]: seems like it's not necessary to call this function on each cycle's iteration. Let's move it above

File: js/ui/calendar/ui.calendar.base_view.js [hunk 3]
```diff
@@ -152,6 +160,26 @@ const BaseView = Widget.inherit({
...
+        cell.className = 'dx-calendar-week-number-cell';
+        cell.innerText = weekNumber;
+
+        if(rtlEnabled) {
+            rowData.row.append(cell);
+        } else {
+            rowData.row.prepend(cell);
+        }
+        this.setAria({
+            'role': 'gridcell'
```
Comment (ksercs) [inline js/ui/calendar/ui.calendar.base_view.js]: Don't we need a `label` here?

File: js/ui/calendar/ui.calendar.base_view.js [hunk 4]
```diff
@@ -152,6 +160,26 @@ const BaseView = Widget.inherit({
...
     },
 
+    _renderWeekNumberCell: function(rowData) {
+        const { firstDayOfWeek, rtlEnabled } = this.option();
+        const weekNumber = coreDateUtils.getWeekNumber(rowData.prevCellDate, firstDayOfWeek);
+
+        const cell = domAdapter.createElement('td');
+        const $cell = $(cell);
+
+        cell.className = 'dx-calendar-week-number-cell';
```
Comment (ksercs) [inline js/ui/calendar/ui.calendar.base_view.js]: let's move it to a constant?

File: js/core/utils/date.js [hunk 5]
```diff
@@ -502,6 +502,15 @@ function getFirstWeekDate(date, firstDayOfWeek) {
     return result;
 }
 
+function getWeekNumber(date, firstDayOfWeekIndex = 0) {
+    const firstDay = new Date(date.getFullYear(), 0, 1);
```
Comment (ksercs) [inline js/core/utils/date.js]:
```suggestion
    const yearFirstDay = new Date(date.getFullYear(), 0, 1);
```

File: testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js [hunk 6]
```diff
@@ -1678,6 +1680,84 @@ QUnit.module('Options', {
...
+            this.checkColumnCount(8);
+        });
+
+        QUnit.test('table should not have additional column if showWeekNumbers=false', function() {
+            this.reinit({ showWeekNumbers: false });
+
+            this.checkColumnCount(7);
+        });
+
+        QUnit.test('table should be rerendred with additional column after runtime chage of showWeekNumbers', function(assert) {
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js]:
```suggestion
        QUnit.test('table should be rerendered with additional column after runtime change of showWeekNumbers', function(assert) {
```

File: testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js [hunk 7]
```diff
@@ -1678,6 +1680,84 @@ QUnit.module('Options', {
 
         assert.deepEqual(today, result, 'today date is correct');
     });
+
+    QUnit.module('showWeekNumbers', {
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js]: I think this module should be moved to `calendar.markup.tests` for testing in SSR

File: testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js [hunk 8]
```diff
@@ -1678,6 +1680,84 @@ QUnit.module('Options', {
...
+            const $lastBodyRowCell = this.$firstBodyRow.children().eq(7);
+
+            assert.ok($lastBodyRowCell.hasClass(CALENDAR_WEEK_NUMBER_CELL_CLASS));
+        });
+
+        QUnit.test('calendar with zoomLevel!=="month" and showWeekNumbers=true should not have additional column', function(assert) {
+            this.reinit({ showWeekNumbers: true, zoomLevel: 'year' });
+            this.cacheTableElements();
+
+            assert.ok(true);
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js]: seems like you don't use TDD :)

File: js/core/utils/date.js [hunk 9]
```diff
@@ -502,6 +502,15 @@ function getFirstWeekDate(date, firstDayOfWeek) {
     return result;
 }
 
+function getWeekNumber(date, firstDayOfWeekIndex = 0) {
```
Comment (ksercs) [inline js/core/utils/date.js:505]: We should add tests for this method to `utils.date.tests.js`

File: testing/testcafe/tests/editors/calendar/calendar.ts [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+import { restoreBrowserSize } from '../../../helpers/restoreBrowserSize';
+
+fixture`Calendar`
+  .page(url(__dirname, '../../container.html'))
+  .afterEach(async () => {
+    await changeTheme('generic.light');
+  });
+
+const themes = ['generic.light', 'generic.light.compact', 'material.blue.light', 'material.blue.light.compact'];
+themes.forEach((theme) => {
```
Comment (ksercs) [inline testing/testcafe/tests/editors/calendar/calendar.ts:13]: What about test for `rtlEnabled=true`?

File: testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js [hunk 11]
```diff
@@ -1678,6 +1680,84 @@ QUnit.module('Options', {
...
+        });
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js]: Let's test columns count after `zoomLevel` prop change

File: js/core/utils/date.js [hunk 12]
```diff
@@ -502,6 +502,15 @@ function getFirstWeekDate(date, firstDayOfWeek) {
...
+    const firstDay = new Date(date.getFullYear(), 0, 1);
+    const differenceInDays = dateUtils.getDatesInterval(firstDay, date, 'day');
+    const weekOffset = (7 + new Date(Date.UTC(date.getFullYear(), 0, 1)).getUTCDay() - firstDayOfWeekIndex) % 7;
```
Comment (ksercs) [inline js/core/utils/date.js]: Uh, I don't manage to understand this formula. Can u explain it?
Seems like we should introduce `getFirstDayOfFirstWeekInYear` function. Then we can just set 
```javascript
differenceInDays = dateUtils.getDatesInterval(firstDayOfFirstWeekInYear, date, 'day');;
week = Math.floor((differenceInDays - 1) / 7) + 1;
```

File: testing/testcafe/tests/editors/calendar/calendar.ts [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+  test(`Calendar with showWeekNumbers rendered correct (${theme})`, async (t) => {
```
Comment (ksercs) [inline testing/testcafe/tests/editors/calendar/calendar.ts:14]: Let's also add a corner case when the last row is out of this year (for example, December, 2022). Is the behavior correct in this case?

File: js/ui/calendar/ui.calendar.base_view.js [hunk 14]
```diff
@@ -152,6 +160,26 @@ const BaseView = Widget.inherit({
         params.cellDate = this._getNextCellData(cellDate);
     },
 
+    _renderWeekNumberCell: function(rowData) {
```
Comment (ksercs) [inline js/ui/calendar/ui.calendar.base_view.js:166]: How [cellTemplate](https://js.devexpress.com/Documentation/ApiReference/UI_Components/dxCalendar/Configuration/#cellTemplate) option should work for this cells? What parameters it accepts? `data: undefined, weekNumber: number`? 

We should write tests for this and update documentation.

File: testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js [hunk 15]
```diff
@@ -1678,6 +1680,84 @@ QUnit.module('Options', {
...
+    });
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js]: What about accessibility? Did u test it?

We should add tests.

**Review by ksercs — CHANGES_REQUESTED** [review]

File: js/ui/calendar/ui.calendar.base_view.js (hunk 2 above)
Comment (Zedwag) [reply js/ui/calendar/ui.calendar.base_view.js]: ok

File: js/ui/calendar/ui.calendar.base_view.js (hunk 4 above)
Comment (Zedwag) [reply js/ui/calendar/ui.calendar.base_view.js]: ok

File: js/core/utils/date.js (hunk 5 above)
Comment (Zedwag) [reply js/core/utils/date.js]: ok

File: testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js (hunk 6 above)
Comment (Zedwag) [reply testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js]: done

File: testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js (hunk 8 above)
Comment (Zedwag) [reply testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js]: u got me, sherlock

File: testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js (hunk 7 above)
Comment (Zedwag) [reply testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js]: ok

File: testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js (hunk 8 above)
Comment (Zedwag) [reply testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js]: fixed

File: testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js (hunk 11 above)
Comment (Zedwag) [reply testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js]: done

File: js/ui/calendar/ui.calendar.base_view.js (hunk 14 above)
Comment (Zedwag) [reply js/ui/calendar/ui.calendar.base_view.js:166]: should it work for this cells?

File: js/core/utils/date.js (hunk 12 above)
Comment (Zedwag) [reply js/core/utils/date.js]: nope, that will not work with different `firstDayOfWeek`

Comment (Zedwag) [reply js/core/utils/date.js]: also it does not work for some simple cases with default `firstDayOfWeek`
image

File: testing/testcafe/tests/editors/calendar/calendar.ts (hunk 13 above)
Comment (Zedwag) [reply testing/testcafe/tests/editors/calendar/calendar.ts:14]: added

File: testing/testcafe/tests/editors/calendar/calendar.ts (hunk 10 above)
Comment (Zedwag) [reply testing/testcafe/tests/editors/calendar/calendar.ts:13]: added

File: js/core/utils/date.js (hunk 12 above)
Comment (Zedwag) [reply js/core/utils/date.js]: formula is simplified

File: js/core/utils/date.js [hunk 16]
```diff
@@ -502,6 +502,14 @@ function getFirstWeekDate(date, firstDayOfWeek) {
     return result;
 }
 
+function getWeekNumber(date, firstDayOfWeekIndex = 0) {
+    const yearFirstDay = new Date(date.getFullYear(), 0, 1);
```
Comment (EugeniyKiyashko) [inline js/core/utils/date.js:506]: Will be it correct with various time zones?

File: testing/testcafe/tests/editors/calendar/calendar.ts [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+import { restoreBrowserSize } from '../../../helpers/restoreBrowserSize';
+
+fixture`Calendar`
+  .page(url(__dirname, '../../container.html'))
+  .afterEach(async () => {
+    await changeTheme('generic.light');
+  });
+
+const themes = ['generic.light', 'generic.light.compact', 'material.blue.light', 'material.blue.light.compact'];
+themes.forEach((theme) => {
```
Comment (EugeniyKiyashko) [inline testing/testcafe/tests/editors/calendar/calendar.ts:13]: Let's add the tests for rtl mode

Comment (EugeniyKiyashko) [reply testing/testcafe/tests/editors/calendar/calendar.ts:13]: added

File: testing/testcafe/tests/editors/calendar/calendar.ts [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+  test(`Calendar with showWeekNumbers rendered correct (${theme})`, async (t) => {
+    await t.expect(await compareScreenshot(t, `calendar-with-showWeekNumbers-theme=${theme.replace(/\./g, '-')}.png`)).ok();
+  }).before(async (t) => {
+    await t.resizeWindow(300, 300);
```
Comment (EugeniyKiyashko) [inline testing/testcafe/tests/editors/calendar/calendar.ts]: I think we can avoid resizeWindow() use and take a screenshot of '#container' element directly

Comment (EugeniyKiyashko) [reply testing/testcafe/tests/editors/calendar/calendar.ts]: In all tests

File: scss/widgets/generic/calendar/_index.scss [hunk 19]
```diff
@@ -194,6 +194,19 @@ $generic-calendar-height: $generic-calendar-body-height + $generic-calendar-navi
...
+      color: $calendar-cell-selected-bg;
+    }
+
+    .dx-calendar-week-number-cell {
+      text-align: center;
+      padding: 1px $generic-calendar-cell-horizontal-padding 2px;
+      color: $calendar-cell-selected-bg;
+      font-size: $generic-calendar-week-number-cell-font-size;
+      width: 39px;
+      cursor: default;
```
Comment (EugeniyKiyashko) [inline scss/widgets/generic/calendar/_index.scss]: Can we extract common styles such text-align, cursor to common files?

File: js/ui/calendar/ui.calendar.base_view.js (hunk 14 above)
Comment (EugeniyKiyashko) [reply js/ui/calendar/ui.calendar.base_view.js:166]: I think we should have a capability to customize these cells

File: js/ui/calendar/ui.calendar.base_view.js [hunk 20]
```diff
@@ -111,6 +117,11 @@ const BaseView = Widget.inherit({
             'append';
     },
 
+    _shouldRenderWeekColumn: function() {
+        const { showWeekNumbers, zoomLevel } = this.option();
+        return showWeekNumbers && zoomLevel === 'month';
```
Comment (EugeniyKiyashko) [inline js/ui/calendar/ui.calendar.base_view.js:122]: zoomLevel === 'month' -> ZOOM_LEVEL.MONTH

File: js/core/utils/date.js (hunk 16 above)
Comment (Zedwag) [reply js/core/utils/date.js:506]: As far as `new Date` takes current timezone, I don't think there should be any problems with timezone

File: js/ui/calendar/ui.calendar.base_view.js (hunk 20 above)
Comment (Zedwag) [reply js/ui/calendar/ui.calendar.base_view.js:122]: there is no ZOOM_LEVEL object in base_view file, only in ui.calendar.js. And it's the only place where we check zoomLevel in base view, so I see no point to add this const here

File: testing/testcafe/tests/editors/calendar/calendar.ts (hunk 17 above)
Comment (Zedwag) [reply testing/testcafe/tests/editors/calendar/calendar.ts:13]: I added one next to these, do u think we need test rtl in all themes?

File: scss/widgets/generic/calendar/_index.scss (hunk 19 above)
Comment (Zedwag) [reply scss/widgets/generic/calendar/_index.scss]: done

File: js/ui/calendar/ui.calendar.base_view.js (hunk 14 above)
Comment (Zedwag) [reply js/ui/calendar/ui.calendar.base_view.js:166]: implemented

File: js/ui/calendar/ui.calendar.base_view.js (hunk 3 above)
Comment (Zedwag) [reply js/ui/calendar/ui.calendar.base_view.js]: I added aria-label and aria-readonly

File: js/ui/calendar/ui.calendar.base_view.js (hunk 14 above)
Comment (Zedwag) [reply js/ui/calendar/ui.calendar.base_view.js:166]: thrown to the trash can and started again

File: testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js (hunk 15 above)
Comment (Zedwag) [reply testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js]: added test on aria-label and role of week number cells

File: testing/testcafe/tests/editors/calendar/calendar.ts (hunk 18 above)
Comment (Zedwag) [reply testing/testcafe/tests/editors/calendar/calendar.ts]: done

File: js/ui/calendar/ui.calendar.base_view.js (hunk 14 above)
Comment (Zedwag) [reply js/ui/calendar/ui.calendar.base_view.js:166]: nope seem good, implemented and added a quinit + one screenshot test for templates

File: js/ui/calendar/ui.calendar.base_view.js [hunk 21]
```diff
@@ -152,6 +163,32 @@ const BaseView = Widget.inherit({
...
+        }
+
+        if(rtlEnabled) {
+            rowData.row.append(cell);
+        } else {
+            rowData.row.prepend(cell);
+        }
+        this.setAria({
+            'role': 'gridcell',
+            'label': `Week ${weekNumber}`,
```
Comment (EugeniyKiyashko) [inline js/ui/calendar/ui.calendar.base_view.js:188]: Do we need to localize the label text?

File: js/ui/calendar/ui.calendar.base_view.js (hunk 14 above)
Comment (EugeniyKiyashko) [reply js/ui/calendar/ui.calendar.base_view.js:166]: Let's add the commit hash please to see the changes where is a implementation

File: js/ui/calendar/ui.calendar.base_view.js (hunk 20 above)
Comment (EugeniyKiyashko) [reply js/ui/calendar/ui.calendar.base_view.js:122]: May be it should shared const variable in common files between the files it is used? Or can we import this const from file where it is declared?

File: testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js [hunk 22]
```diff
@@ -1488,6 +1489,21 @@ QUnit.module('Options', {
         assert.strictEqual(cellContent, 'Custom template', 'cell content is correct after cellTemplate runtime change');
     });
 
+    QUnit.test('cellTemplate is rendered fow week cell', function(assert) {
+        this.calendar.option({
```
Comment (EugeniyKiyashko) [inline testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js:1493]: I want to know which cellData is for 'week' cells? Can we check it?

File: testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js [hunk 23]
```diff
@@ -3623,6 +3639,22 @@ QUnit.module('Aria accessibility', {
             });
         });
     });
+
+    [
+        { attr: 'aria-label', value: 'Week 1' },
```
Comment (EugeniyKiyashko) [inline testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js:3644]: Is it right that we can customize attributes for `week` cells?

File: testing/testcafe/tests/editors/calendar/calendar.ts [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+test('Calendar with showWeekNumbers rendered correct for last week of year value', async (t) => {
+  const { takeScreenshot, compareResults } = createScreenshotsComparer(t);
+
+  await t
+    .expect(await takeScreenshot('calendar-with-showWeekNumbers-last-week.png', '#container'))
+    .ok()
+    .expect(compareResults.isValid())
+    .ok(compareResults.errorMessages());
+}).before(async () => {
+  await createWidget('dxCalendar', {
```
Comment (EugeniyKiyashko) [inline testing/testcafe/tests/editors/calendar/calendar.ts:41]: Why the `last week` number is `1` but not `54` in your scenario. I think that logic to calculate week number works incorrectly, we need double check it.

Why? 
In Google Calendar for december of 2021 year I see that the last week number is 52.
image

Comment (EugeniyKiyashko) [reply testing/testcafe/tests/editors/calendar/calendar.ts:41]: And you variant 
image

File: testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js (hunk 22 above)
Comment (Zedwag) [reply testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js:1493]: Yes, week number cells have `cellIndex: -1` in screenshot test it's shown how to make week cells bold and others cell italic

File: testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js (hunk 23 above)
Comment (Zedwag) [reply testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js:3644]: No, it's just accessibility test

File: js/ui/calendar/ui.calendar.base_view.js (hunk 14 above)
Comment (Zedwag) [reply js/ui/calendar/ui.calendar.base_view.js:166]: no hash, parts in two pr (3 and 4), implemented here: https://github.com/DevExpress/DevExtreme/pull/22674/files#diff-5efeafa614b53cfce71ef52ae5b44108e26958f71e4efb4d908a0bd81689f029R176 and here https://github.com/DevExpress/DevExtreme/pull/22674/files#diff-5efeafa614b53cfce71ef52ae5b44108e26958f71e4efb4d908a0bd81689f029R211

File: js/ui/calendar/ui.calendar.base_view.js (hunk 21 above)
Comment (Zedwag) [reply js/ui/calendar/ui.calendar.base_view.js:188]: hm, idk actually

File: testing/testcafe/tests/editors/calendar/calendar.ts (hunk 24 above)
Comment (Zedwag) [reply testing/testcafe/tests/editors/calendar/calendar.ts:41]: it's because of american system of counting week numbers

Comment (EugeniyKiyashko) [reply testing/testcafe/tests/editors/calendar/calendar.ts:41]: Ok, let's create a howTo ticket to demonstrate the approved solution how we can change the system using cellTemplate

File: js/ui/calendar/ui.calendar.base_view.js (hunk 21 above)
Comment (Zedwag) [reply js/ui/calendar/ui.calendar.base_view.js:188]: desided not to localize this

File: testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js (hunk 23 above)
Comment (Zedwag) [reply testing/tests/DevExpress.ui.widgets.editors/calendar.tests.js:3644]: Yep wia cellTemplate

File: js/ui/calendar/ui.calendar.base_view.js [hunk 25]
```diff
@@ -171,12 +208,12 @@ const BaseView = Widget.inherit({
     },
 
     _prepareCellTemplateData: function(cellDate, cellIndex, $cell) {
+        const text = cellDate instanceof Date ? this._getCellText(cellDate) : cellDate;
+        const date = cellDate instanceof Date ? cellDate : undefined;
```
Comment (ksercs) [inline js/ui/calendar/ui.calendar.base_view.js]: `const isDateCell =cellDate instanceof Date;`
Let's reduce copypaste

Comment (Zedwag) [reply js/ui/calendar/ui.calendar.base_view.js]: done

**Review by ksercs — APPROVED** [review]

**Review by EugeniyKiyashko — APPROVED** [review]
